### PR TITLE
libct: add/use configs.HasHook

### PR DESCRIPTION
**Pull request by kolyshkin**
1. libct: add/use configs.HasHook
    
    This allows to omit a call to c.currentOCIState (which can be somewhat
    costly when there are many annotations) when the hooks of a given kind
    won't be run.

2. libct: don't pass SpecState to init unless needed
    
    SpecState field of initConfig is only needed to run hooks that are
    executed inside a container -- namely CreateContainer and
    StartContainer.
    
    If these hooks are not configured, there is no need to fill, marshal and
    unmarshal SpecState.

### CODE REVIEW

File: libcontainer/process_linux.go
Comment (cyphar) [inline libcontainer/process_linux.go:832]: Minor nit: While there is only one caller of this and the semantics make sense for that caller, I think it'd be a bit easier to follow if the `if HasHook` check was in the caller and this method stayed as `updateSepcState`. As it stands, "maybe" doesn't tell me that much when I'm looking at the caller's code.

Comment (kolyshkin) [reply libcontainer/process_linux.go]: I agree.

Even better, we can inline it since it's trivial and only has a single user.

Please see updated.

**Review by lifubang — APPROVED** [review]

**Comment by kolyshkin** [comment]
Rebased, resolving the conflict after #4312 merge. PTAL @opencontainers/runc-maintainers, this is an easy to review one.

**Review by AkihiroSuda — APPROVED** [review]

File: libcontainer/configs/config.go
Comment (lifubang) [inline libcontainer/configs/config.go:335]: I think we don’t have libct API document at this time. Otherwise this useful API should be mentioned in the doc.

**Pull request merged**
